### DISCUSSION TITLE
"-channel RGB -compose multiply -composite" works differently in ImageMagick 7 than ImageMagick 6?

### DISCUSSION THREAD

**Discussion by aronyu79** [ImageMagick/ImageMagick#4251] [Help]
I have two source images:
**extract-0**
<img src="https://user-images.githubusercontent.com/12458094/133926350-ee2baab9-74b9-46c5-8dc5-a0dc560a42c3.png" width="100" height="100">
**extract-1**
<img src="https://user-images.githubusercontent.com/12458094/133926349-3d7dd54c-298b-4d84-a644-09c31b26aee2.png" width="100" height="100">

With ImageMagick 6.9.6-0 Q16 x86 2016-10-02
I could premultiply alpha e.g. like this:
`convert6 .\extract-0.png .\extract-1.png -channel RGB -compose multiply -composite im6compositedRGBnosync.png`

Yielding this image:
<img src="https://user-images.githubusercontent.com/12458094/133926478-4eea7303-300c-4694-9bf3-196024836cc9.png" width="100" height="100">

Same command on ImageMagick 7.1.0-7 Q16-HDRI x64
`convert7.\extract-0.png .\extract-1.png  -channel RGB -compose multiply -composite im7compositedRGBnosync.png`
Results in this image:
<img src="https://user-images.githubusercontent.com/12458094/133926523-b83e1d4a-ee90-4aa1-9f41-e463c65882f0.png" width="100" height="100">
The green and blue pixel values are wrong.

As a sidenote I noticed this is the same pixel values as running with sync on ImageMagick6
`convert6 .\extract-0.png .\extract-1.png -channel RGB,sync -compose multiply -composite im6compositedRGBsync.png`
<img src="https://user-images.githubusercontent.com/12458094/133926453-e346959e-4de4-4bb8-bc38-719da40a615e.png" width="100" height="100">
Although the alpha is different here.

I am wondering if it is a bug that ImageMagick7 ignores that the sync flag is removed when multiplying?
Or why is these parameters different on ImageMagick7
`.\input-0.png .\input-1.png -channel RGB -compose multiply -composite output.png`

**Comment by fmw42** [comment]
Your syntax is wrong especially for ImageMagick 7. Proper syntax for IM 6 and IM 7 is to read the input image first. So

`convert input0 input1 -channel rgb -compose multiply -composite output`

For IM 7, use magick not convert.

I am not sure if that will make a difference but try that.

**Reply by aronyu79** [reply]
For IM7, there are symbolic links to to magick which actually does magick convert. For Windows there is still a binary in the portable package called convert.exe. I am running on Windows.

I realized I had the params in the wrong order and changed the post. I believe you might have read it before I changed. I tried the correct order (i.e. the one you suggested) but it has the same result.

**Comment by snibgo** [comment]
IM uses either v6 syntax or v7 syntax. A "magick" command will always use v7, a "convert" command will always use v6. This is true even if one is a symbolic link to another, because the program uses argv[0] to decide. "magick convert" always uses v6 syntax.

I recommend that if you use v7, you always use "magick".

That point aside, I think you have correctly identified a difference in v6 and v7. I think the v7 result is wrong, at least partly. For the result alpha (ignoring complexities around "sync"), I think the calculation should be:
```
Da'  =         1 * Sa * Da  + 1 * Sa * (1-Da)   + 1 * Da * (1-Sa)

```
... where Da' is the output alpha, Sa is the source alpha, Da is the destination alpha, given the command is:
```
magick destination source -compose Multiply -composite output
```
So:
```
Da = 128/255 = 0.5020
Sa = 1
```

So:
```
Da'  =         1 * 1 * 0.5020  + 1 * 1 * (1-0.5020)   + 1 * 0.5020 * (1-1)
     =         0.5020  + (1-0.5020) + 0
     = 1
```

So the output alpha should be 1, fully opaque. V6.9.9-50 gives that result when we use "-channel RGB,sync" or no channel setting at all. V7.1.0-4 never gives that result.

So v7.1.0-4 seems to have a bug.

**Reply by aronyu79** [reply]
Ok thanks for clarifying about magick vs convert. (tho convert 7 is not fully compatible with the convert 6 commandline)

I forgot to tag which image was which above (I've edited the post now).

I am not sure what the equation should be. With/without sync on ImageMagick 6 I get different alpha values (and color values). Without sync I am getting 0.5/128 which is what I am expecting (because A-channel isn't multiplied?), with sync it is fully opaque.

The alpha value seem ok with ImageMagick7 (since we are not multiplying it). My main concern are the color values. But maybe the same bug?

Do you think I should file a bug report somewhere?

**Comment by snibgo** [comment]
The formula comes from my page [Compose operators](http://im.snibgo.com/compose.htm), which comes from the SVG Compositing Specification https://www.w3.org/TR/SVGCompositing/ . As always, I welcome corrections.

When there is no alpha, "-compose Multiply -composite" should simply multiply each channel. This seems correct for IM v6 and v7. The problems arise when we have transparency.

"Sync" is an IM concept that is not in the SVG spec. When IM uses sync, it should comply with the SVG spec. IM documentation is a bit vague about what happens when sync is not used. See https://legacy.imagemagick.org/Usage/compose/#sync_flag

For the colour channels, the SVG formula is:
```
Dca' = f(Sc, Dc) * Sa * Da  + Y * Sca * (1-Da)  + Z * Dca * (1-Sa)
```
... where Dca' is the output colour channel multiplied by the output alpha, f(Sc,Dc) is Sc*Dc, Sca and Dca are the source and destination colour channels multiplied by their alphas, and Y=Z=1. As above, Sa and Da are the source and destination input alphas.
```
Dca' = Sc*Dc * Sa * Da    + 1 * Sca * (1-Da)      + 1 * Dca * (1-Sa)
     = Sc*Dc * 1 * 0.5020 + 1 * Sca * (1-0.5020)  + 1 * Dca * (1-1)
     = Sc*Dc * 0.5020     + Sca * 0.4980          + 0
     = Sc*Dc * 0.5020     + Sc  * 0.4980
```
For example, the destination and source green channels are:
```
Dc = 110/255 = 0.4314
Sc = 128/255 = 0.5020
```
So:
```
Dca' = 0.5020*0.4314 * 0.5020      + 0.5020  * 0.4980
     = 0.3587
```
Dca' is the output colour multipled by its alpha. The output alpha should be 1:
```
Dc'  = Dca' / Da'
     = 0.3587 / 1
     = 0.3587
```

My basic test script for this is:
```
magick ^
  ex0.png ex1.png ^
  -depth 16 ^
  -crop 1x1+0+0 +repage ^
  +write txt: ^
  -compose multiply -composite ^
  +write txt: ^
  x7.png
```
... where ex0.png and ex1.png are your input images. I ignore the output image, and care only about the text:
```
# ImageMagick pixel enumeration: 1,1,65535,srgba
0,0: (63479,28270,48059,32896)  #F7F76E6EBBBB8080  srgba(247,110,187,0.501961)
# ImageMagick pixel enumeration: 1,1,65535,gray
0,0: (32896)  #808080808080  gray(128)
# ImageMagick pixel enumeration: 1,1,65535,srgba
0,0: (32378,23506.5,28492.7,32896)  #7E7A5BD36F4D8080  srgba(49.4056%,35.8687%,43.477%,0.501961)
```
For this example, the v7 green channel is correct, but the alpha is not.

I don't think we need to create an "issue" (at https://github.com/ImageMagick/ImageMagick/issues ) because the developers keep an eye on this forum.

**Reply by aronyu79** [reply]
I will take some time to go thru your post! Thanks for all the info and detailed response! :pray:
